### ISSUE TITLE
bac_metric gives negative values for pipelines that perform worse than chance

### ISSUE THREAD

**Issue by rhiever** [automl/auto-sklearn#105]
First off I want to say: I'm glad that you implemented balanced accuracy! For a short while I thought I was crazy for implementing it in TPOT because I couldn't find this metric in many other places, including sklearn.

Anyway, I've been using auto-sklearn with balanced accuracy: https://github.com/automl/auto-sklearn/blob/master/autosklearn/metrics/classification_metrics.py#L117

and I noticed that the balanced accuracy "scales" the balanced accuracy to [0.0, 1.0], where 0.0 is "as good as chance." However, because of the scaling procedure, it's actually possible for the balanced accuracy to return negative values when pipelines/models perform worse than chance. I've confirmed this with one difficult data set that AutoML methods typically achieve worse-than-random accuracy on.

IMO balanced accuracy should not scale the final score: 0.5 should represent "as good as chance," such that there is room for worse-than-chance models can have an even lower non-negative score.

**Comment by mfeurer** [comment]
Balanced accuracy is in auto-sklearn because it was a central metric in the ChaLearn AutoML challenge. Nevertheless, it does a good job and we will certainly keep it. Regarding sklearn, isn't macro-averaged precision the same thing?

Currently, we're planning to replace all metric code that we've taken from the ChaLearn sample submission code by the scikit-learn code and there is a PR that does a first step towards that: #66

**Comment by rhiever** [comment]
IIRC precision only takes into account when the classifier accurately identifies a class, so as an easy example, a classifier could get 100% precision on one class by always guessing that class. However, it's also important to factor in that the classifier *also* correctly identified when it *wasn't* that class. Hence why we calculate accuracy on a per-class basis (with a one-vs-all scheme) then average the accuracies. To my understanding, that is what the current balanced accuracy function does.

I described this to the sklearn devs and they didn't seem to think that it was a properly formulated metric in the multiclass case. I'm not 100% sure why.

**Comment by mfeurer** [comment]
Hm, I don't think the code in auto-sklearn calculates the class-wise accuracies. It calculates the total number of true positives divided by the total number of positive samples. This might be the same thing as you describe, because classifying a sample wrong is not a true positive and not counted in the numerator of the expression.

**Comment by rhiever** [comment]
It calculates both TPR and TNR: https://github.com/automl/auto-sklearn/blob/master/autosklearn/metrics/classification_metrics.py#L185

**Comment by mfeurer** [comment]
The branch you refer to is only used in binary and (binary) multilabel classification, not in multiclass classification.

**Comment by rhiever** [comment]
Good catch! That's interesting that TNR is dropped for the multiclass case. I'm guessing that's because TNR captures multiple classes in that case.

**Comment by mfeurer** [comment]
Have a look at confusion matrices for 2 class and 3 class problems:

```
TP | FP
-------
FN | TN

TP| ?| ?
--------
? |TP|?
--------
? | ?|TP
```

In the three-class case there is no negative. Imagine the same thing for the two-class case:

```
TP | ?
------
? | TP
```

it's not about true or false, but only about predicting correctly.

**Comment by rhiever** [comment]
It's a good idea to integrate the sklearn metrics then. 👍
